src/Gui/ImageView.cpp

```diff
@@ -0,0 +1,369 @@
+            setCursor(QCursor(Qt::ClosedHandCursor));
+            startDrag();
+            dragPos = event->pos();
+        }
+    }
+}
+
+void ImageView::mouseReleaseEvent(QMouseEvent* event)
+{
+    if (!(event->buttons() & Qt::MiddleButton)) {
+        if (isDragging()) {
+            stopDrag();
+            unsetCursor();
+        }
+    }
+}
+
+void ImageView::mouseMoveEvent(QMouseEvent* event)
+{
+    if (isDragging()) {
+        QScrollBar* hBar = scrollArea->horizontalScrollBar();
+        QScrollBar* vBar = scrollArea->verticalScrollBar();
+        QPoint delta = event->pos() - dragPos;
+        hBar->setValue(hBar->value() + (isRightToLeft() ? delta.x() : -delta.x()));
+        vBar->setValue(vBar->value() - delta.y());
+        dragPos = event->pos();
+    }
+}
+
+void ImageView::dropEvent(QDropEvent* event)
+{
+    const QMimeData* data = event->mimeData();
+    if (data->hasUrls()) {
+        loadImageFromUrl(data->urls());
+    }
+    else {
+        MDIView::dropEvent(event);
+    }
+}
+
+void ImageView::dragEnterEvent(QDragEnterEvent* event)
+{
+    const QMimeData* data = event->mimeData();
+    if (data->hasUrls()) {
+        event->accept();
+    }
+    else {
+        event->ignore();
+    }
+}
+
+bool ImageView::isImageFormat(const QFileInfo& fileInfo) const
+{
+    QString ext = fileInfo.suffix().toLower();
+    QByteArray suffix = ext.toLatin1();
+    QList<QByteArray> supportedFormats = QImageReader::supportedImageFormats();
+    auto it = std::find_if(supportedFormats.begin(), supportedFormats.end(), [suffix](const QByteArray& image) {
+        return (image == suffix);
+    });
+
+    return (it != supportedFormats.end());
+}
+
+void ImageView::loadImageFromUrl(const QList<QUrl>& urls)
+{
+    if (urls.isEmpty())
+        return;
+    QUrl url = urls.first();
+    QFileInfo info(url.toLocalFile());
+    if (info.exists() && info.isFile()) {
+        if (isImageFormat(info)) {
+            loadFile(info.absoluteFilePath());
+        }
+    }
+}
+
+void ImageView::pasteImage()
+{
+    QImage image = imageFromClipboard();
+    if (!image.isNull()) {
+        setImage(image);
+    }
+}
+
+bool ImageView::canPasteImage() const
+{
+    return !imageFromClipboard().isNull();
+}
+
+QImage ImageView::imageFromClipboard() const
+{
+    QImage image;
+    if (const QMimeData *mimeData = QApplication::clipboard()->mimeData()) {
+        if (mimeData->hasImage()) {
+            image = qvariant_cast<QImage>(mimeData->imageData());
+        }
+    }
+
+    return image;
+}
+
+void ImageView::print(QPrinter* printer)
+{
+    QPainter painter(printer);
+    QPixmap pixmap = QPixmap::fromImage(rawImage);
+    QRect rect = painter.viewport();
+    QSize size = pixmap.size();
+    size.scale(rect.size(), Qt::KeepAspectRatio);
+    painter.setViewport(rect.x(), rect.y(), size.width(), size.height());
+    painter.setWindow(pixmap.rect());
+    painter.drawPixmap(0, 0, pixmap);
+}
+
+bool ImageView::onMsg(const char* pMsg,const char**)
```

Comment:
What is this second parameter? clang-tidy complains about its missing name.

Reply:
onMsg() can use the second parameter to report back a string to the calling instance. We have exactly one use case where we do this: View3DInventor::onMsg. It's used in StdViewDockUndockFullscreen::activated

Reply:
Added a named argument